src/relax/op/nn/nn.cc

```diff
@@ -68,5 +60,338 @@ TVM_REGISTER_OP("relax.nn.softmax")
     .set_attrs_type<SoftmaxAttrs>()
     .set_attr<FInferStructInfo>("FInferStructInfo", InferStructInfoSoftmax);
 
+bool NormCheckDtypeAndShape(const Call& call, const BlockBuilder& ctx,
+                            const Array<TensorStructInfo>& input_sinfo,
+                            const Array<String>& input_names, Array<Integer> axes,
+                            const String& op_name) {
+  TensorStructInfo data_sinfo = input_sinfo[0];
+  int n_input = input_sinfo.size();
+
+  std::vector<int> axes_non_neg;
+  if (!data_sinfo->IsUnknownNdim()) {
+    axes_non_neg = CheckAxesInRangeNonRepetitive(call, ctx, data_sinfo->ndim, axes, op_name);
+  }
+  int n_axis = axes.size();
+
+  for (int i = 1; i < n_input; ++i) {
+    if (input_sinfo[i]->dtype != data_sinfo->dtype) {
+      ctx->ReportFatal(Diagnostic::Error(call)
+                       << op_name
+                       << " requires all the input tensors to have the same dtype. However, the "
+                       << input_names[i] << " has dtype " << input_sinfo[i]->dtype
+                       << " which is other than the input data's dtype " << data_sinfo->dtype);
+    } else if (input_sinfo[i]->ndim != n_axis) {
+      ctx->ReportFatal(Diagnostic::Error(call)
+                       << op_name << " requires the input " << input_names[i]
+                       << " to have as many dimensions as the length of input axes. However, the "
+                          "given one has ndim "
+                       << input_sinfo[i]->ndim << ", which is other than the length of axes "
+                       << n_axis);
+    }
+  }
+
+  std::vector<Array<PrimExpr>> axis_lengths;
```

Comment:
Shall we use `Array<Array<PrimExpr>>` here?

Reply:
No.

For one reason, constructing an Array by `push_back` is more time-consuming than constructing a std::vector in the same way, since each `push_back` to an Array does more than just appending a value.

For the other reason, `axis_lengths` is only used internally and won’t be returned and exposed in public API or be passed in FFI.